### PR TITLE
Fixed #26193 -- Fixed urlize to trim multiple trailing punctuation.

### PR DESCRIPTION
**Pull request by jdufresne**
<https://code.djangoproject.com/ticket/26193>

### CODE REVIEW

File: django/utils/html.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 from .html_parser import HTMLParseError, HTMLParser
 
 # Configuration for urlize() function.
-TRAILING_PUNCTUATION = ['.', ',', ':', ';', '.)', '"', '\'', '!']
+TRAILING_PUNCTUATION_RE = re.compile(r'^(.*?)([.,:;!]+)$')
```
Comment (timgraham) [inline django/utils/html.py]: Could you add a comment that explains the regular expression a bit? May not need to be as detailed as https://github.com/django/django/pull/6123 but that's an example of what I have in mind. It seems like some characters that were in `TRAILING_PUNCTUATION` aren't in the regex -- could you explain that to me so I understand it?

File: django/utils/html.py [hunk 2]
```diff
@@ -268,24 +268,40 @@ def unescape(text, trail):
             trail = ''
         return text, unescaped, trail
 
+    def trim_trailing_punctuation(lead, middle, trail):
```
Comment (timgraham) [inline django/utils/html.py]: Some docstrings for these functions would help. The variables aren't named such that their purpose is obvious to me.

File: django/utils/html.py [hunk 3]
```diff
@@ -268,24 +268,40 @@ def unescape(text, trail):
...
+            middle = match.group(1)
+            trail = match.group(2) + trail
+        return lead, middle, trail, bool(match)
+
+    def trim_wrapping_punctuation(lead, middle, trail):
+        trimmed = False
+        for opening, closing in WRAPPING_PUNCTUATION:
+            if middle.startswith(opening):
+                middle = middle[len(opening):]
+                lead = lead + opening
```
Comment (timgraham) [inline django/utils/html.py]: or +=

File: django/utils/html.py [hunk 4]
```diff
@@ -268,24 +268,40 @@ def unescape(text, trail):
...
-                    middle = middle[len(opening):]
-                    lead = lead + opening
-                # Keep parentheses at the end only if they're balanced.
-                if (middle.endswith(closing)
-                        and middle.count(closing) == middle.count(opening) + 1):
-                    middle = middle[:-len(closing)]
-                    trail = closing + trail
+
+            # Continue trimming until middle is unchanged.
+            trim = True
```
Comment (timgraham) [inline django/utils/html.py]: `trimmed_something = True` would be more meaningful.

File: django/utils/html.py [hunk 5]
```diff
@@ -268,24 +268,40 @@ def unescape(text, trail):
...
+            while trim:
+                lead, middle, trail, trimmed1 = trim_trailing_punctuation(lead, middle, trail)
```
Comment (timgraham) [inline django/utils/html.py]: `trimmed_punctuation` would be more meaningful.

File: docs/releases/1.10.txt [hunk 6]
```diff
@@ -324,6 +324,9 @@ Templates
 
 * Added the ``is`` comparison operator to the :ttag:`if` tag.
 
+* The :tfilter:`urlize` template filter now handles multiple trailing
```
Comment (timgraham) [inline docs/releases/1.10.txt]: I guess it doesn't hurt to mention, but personally I wouldn't consider this much of a "feature". :-)

File: django/utils/html.py (hunk 1 above)
Comment (jdufresne) [inline django/utils/html.py]: > It seems like some characters that were in `TRAILING_PUNCTUATION` aren't in the regex -- could you explain that to me so I understand it?

I dropped the closing parenthesis, single quote, and double quote as they are handled by `WRAPPING_PUNCTUATION`. If I missed a detail as to why these need to exist twice, let me know and I'll add a test case for it.

**Comment by jdufresne** [comment]
Updated commit based on feedback.

**Comment by timgraham** [comment]
A few edits: http://dpaste.com/24AC984
I wonder if there's much advantage to the two separate trim functions? Seems like 1 method with two commented sections might be simpler.

**Comment by jdufresne** [comment]
I updated with your edits. Thanks.

> I wonder if there's much advantage to the two separate trim functions? Seems like 1 method with two commented sections might be simpler.

Agreed. I like this better.

I consolidated the two functions down to one and updated the docblock to reflect this. The loop was also moved into the function so the function only needs to be called once.

**Comment by timgraham** [comment]
merged in dec334cb66b3ee59cb82e1bb99a584aa0b9fbbd5, thanks!

**Pull request closed without merging**